File: 799/main.py

class Solution:
    def champagneTower(self, poured: int, query_row: int, query_glass: int) -> float:
        glasses = [[0 for _ in range(x)] for x in range(1, 101)]
        glasses[0][0] = poured

        for i in range(query_row):
            for j in range(len(glasses[i])):
                temp = (glasses[i][j] - 1) / 2.0
                if temp > 0:
                    glasses[i+1][j] += temp
                    glasses[i+1][j+1] += temp

        return glasses[query_row][query_glass] if glasses[query_row][query_glass] <= 1 else 1


if __name__ == "__main__":
    print(Solution().champagneTower(6, 3, 2))

File: 799/main.go
package main

func champagneTower(poured int, query_row int, query_glass int) float64 {
	glasses := make([][]float64, 100)
	for i := range glasses {
		glasses[i] = make([]float64, i+1)
	}
	glasses[0][0] = float64(poured)

	for i, row := range glasses {
		for j, glass := range row {
			t := (glass - 1) / 2
			if t > 0 {
				if i < 100 {
					glasses[i+1][j] += t
					glasses[i+1][j+1] += t
				}
			}
		}
	}
	if glasses[query_row][query_glass] <= 1 {
		return glasses[query_row][query_glass]
	}
	return 1
}
